Development for a MVP of Amber based QM/MM

shaoqx opened the pull request:
interface:
- QMMM arguments in AmberInterface.build_md_step
- convert_stru_to_inpcrd
- yield also pbc_box_shape in AmberMDCRDParser

structure:
- Structure() now carries pbc_box_shape 
- support `same_chain_id_for_9999_solvent` in saving PDB file to avoid bug caused by overflowing residue/chain index
- significantly speed up PrmtopParser

workflow_app:
- a MVP main script for running single point QMMM based on Amber and Gaussian

TODO leftovers for further development:
- Fit AmberMDStep approach into the science API of single_point
- Fit Multiwfn hirshfeld calculation into an API under analysis

shaoqx committed:
Update: fix chain type display

shaoqx committed:
Update: add pbc_box_shape attribute in Structure

shaoqx committed:
Update: parse pbc_box_shape from prmtop

shaoqx committed:
Update: update comments for MDCRD parser

shaoqx committed:
Update: modifed the I/O of MDCRD parser and tested
(add pbc information in the return)

shaoqx committed:
Update: support PBC info in structure_ensemble

shaoqx committed:
Update: reduce from O(N^2) to O(N)

shaoqx committed:
Update: fix the test answer (checked manually)

shaoqx committed:
Update: add is_solvent_chain and test

shaoqx committed:
Update: support using same chain id for solvent
in get_file_str

shaoqx committed:
Update: added convert_stru_to_inpcrd and test

shaoqx committed:
Update: add QMMM options in build_md_step and conf

shaoqx committed:
Update: add alternative names for gaussian

shaoqx committed:
Update:fix and confirm pass the old raw_dict tests

shaoqx committed:
Update: test _parse_md_config_dict_to_raw for qmmm

shaoqx committed:
Update: supported qmmm in _parse_md_config_dict_to
_raw and passed tests

shaoqx committed:
Update:add test for qmmm and fix the bug caused by
the global change in 859b2192bec5215002be7fe53ee8b67d8d2cf481

shaoqx committed:
Update: qmmm md step making pass initial tests

shaoqx committed:
Update: add TODO mark

shaoqx committed:
Update:fix bugs and improve performance of mapping

shaoqx committed:
Update: fix test

shaoqx committed:
Update: left over function for performance

shaoqx committed:
Update: support a new solvent chain threatment
in pdb_io and fixed bug when converting a structure with too many solvent to inpcrd

shaoqx committed:
Update: comment on unfinished

shaoqx committed:
Update: comment on more unfinished

shaoqx committed:
Update: add main script

chrisjurich commented on enzy_htp/_config/gaussian_config.py:
Cool attribute/good idea.

shaoqx replied:
Thanks!

chrisjurich commented on enzy_htp/_interface/amber_interface.py:
```diff
@@ -1021,17 +1090,48 @@ class AmberInterface(BaseInterface):
+    """The mapper that maps keywords of the method to the option int"""
+
+    SQM_METHOD_KEYWORD_MAPPER = {
+        "AM1" : "AM1",
```
AM1 is disgusting... lol

shaoqx replied:
lol

chrisjurich commented on enzy_htp/structure/chain.py:
```diff
@@ -170,6 +170,8 @@ def chain_type(self) -> str:
         if self.has_solvent():
             chain_type.append("solvent")
+        if self.has_residue_cap():
+            chain_type.append("caps")
```
Do you think the `chain_type` should be a `str` or an enum? Not a major thing to think about for now, but consider messing with it in the future

shaoqx replied:
Agree. I made this mainly for supporting the `__str__` of Structure but making it similar to ResidueType is a good idea.

chrisjurich commented on enzy_htp/structure/residue.py:
```diff
@@ -206,7 +206,12 @@ def mainchain_atoms(self) -> List[Atom]:
         return {atom.idx : atom for atom in self.atoms}
-    
+
+    @property
```
great method!

shaoqx replied:
Thanks! Profiling pointed the way for speeding things up. I was very slow lol.

chrisjurich commented on enzy_htp/workflow_app/qmmm_single_point_main.py:
```diff
@@ -0,0 +1,144 @@
+    # 2. do a special iteration of stru_esm that gives paths of inpcrd?
+    jobs = []
+    output_eggs = []
+    for i, stru in enumerate(stru_esm.structures()):
```
I get the impression from this that the SPE for each frame is run in parallel? Great!

shaoqx replied:
Yes!

chrisjurich commented on test/_interface/test_amber_interface.py:
```diff
@@ -722,6 +726,129 @@ def test_parse_md_config_dict_to_raw_wo_cons():
+    assert test_raw_dict == answer_raw_dict
+
+
+def test_parse_md_config_dict_to_raw_qmmm():
```
When run on ACCRE, this test fails, as the `answer_raw_dict['namelists'][2]['method']` is `PBE0`, whereas the `tesst_raw_dict['namelists'][2]['method']` is `pbe1pbe`

shaoqx replied:
Oh right. My mistake

shaoqx replied:
Fixed!

chrisjurich commented on test/_interface/test_amber_interface.py:
```diff
@@ -896,6 +1053,67 @@ def test_amber_md_step_make_job():
+    assert test_md_egg.rst_path == './MD/amber_md_step.rst'
+    assert Path(test_md_egg.prmtop_path) == Path('test/_interface/data//KE_07_R7_S.prmtop').absolute()
+    fs.safe_rmdir(md_step.work_dir)
+    assert False # TODO add more QMMM result assert 
```
Everyting passed up until this point! so good job :)

shaoqx replied:
:)

chrisjurich commented on test/_interface/test_amber_interface.py:
```diff
@@ -955,7 +1172,7 @@ def test_amber_md_step_try_merge_jobs(caplog):
         for test, answer in zip(merged_jobs[0].mimo["commands"], [
                 r'pmemd.cuda -O -i ./MD/amber_md_step_?[0-9]*.in -o ./MD/amber_md_step.out -p .*test/_interface/data//KE_07_R7_S.prmtop -c .*test/_interface/data//KE_07_R7_S.inpcrd -r ./MD/amber_md_step.rst -ref .*test/_interface/data//KE_07_R7_S.inpcrd -x ./MD/amber_md_step.nc ',
-                r'pmemd.cuda -O -i ./MD/amber_md_step_?[0-9]*.in -o ./MD/amber_md_step.out -p .*test/_interface/data//KE_07_R7_S.prmtop -c ./MD/amber_md_step.rst -r ./MD/amber_md_step.rst -ref ./MD/amber_md_step.rst -x ./MD/amber_md_step.nc '
+                r'pmemd.cuda -O -i ./MD/amber_md_step_?[0-9]*.in -o ./MD/amber_md_step.out -p .*test/_interface/data//KE_07_R7_S.prmtop -c ./MD/amber_md_step.inpcrd -r ./MD/amber_md_step.rst -ref ./MD/amber_md_step.inpcrd -x ./MD/amber_md_step.nc '
```
not going to mess with this one for now, with what you said about ACCRE acting up with GPUs.

shaoqx replied:
Right, this is good on old accre and will be good if use AccreR9() for the new one

shaoqx replied:
Oh this one is from a mistake that I batch replaced all "rst" with "inpcrd" in this file

chrisjurich commented on test/_interface/test_amber_interface.py:
```diff
@@ -1404,3 +1621,58 @@ def test_check_prmtop_nc_consistency():
 
     assert result == False
+
+def test_mdcrd_parser_get_coordinates():
```
Passed! congrats!

chrisjurich commented on test/_interface/test_amber_interface.py:
```diff
@@ -1404,3 +1621,58 @@ def test_check_prmtop_nc_consistency():
+        assert coords[-1] == answer_m1
+        assert pbc_box_edges == answer_pbc_edgs
+
+def test_convert_stru_to_inpcrd():
```
Passed! Congrats!

chrisjurich commented on test/_interface/test_amber_interface.py:
```diff
@@ -1404,3 +1621,58 @@ def test_check_prmtop_nc_consistency():
+
+    fs.clean_temp_file_n_dir([test_out_path])
+
+def test_convert_stru_to_inpcrd_too_many_solvent():
```
Passed! Congrats!

chrisjurich commented on test/structure/structure_io/test_prmtop_io.py:
All tests passed for this file! Congrats!

chrisjurich commented on test/structure/test_chain.py:
All tests passed for this file! Congrats!

chrisjurich commented on test/structure/test_structure_ensemble.py:
All tests passed for this file! Congrats!

chrisjurich committed:
made a small change to enzy_htp/_interface/amber_interface.py to fix a possible future dependency error

chrisjurich reviewed (changes requested):
Overall looks very solid! you are implementing an important and big part of molecular mechanics studies, so it is very impressive to have gotten all of this done. The only two things to note are: 
+ a change to importing `Iterable` from `collections.abc` instead of `collections (I have made this change already so it's not a big deal, just something to be aware of
+ Fixing the unit test `test/_interface/test_amber_interface.py:test_parse_md_config_dict_to_raw_qmmm`, which currently errors (see the associated comment for more information)

Once the unit test issue is resolved, will be happy to approve for merge

shaoqx committed:
Bugfix: finally fixed the bug caused by ->
c46434c695ce0e3e058063566045512c87d345f8
and mentioned (wrongly) in ca6e47573aa7dcf87dc7fd91791c23009092621a

shaoqx committed:
Update: fix pbe name in answer

shaoqx commented:
Thank you! Finally fixed the annoying bug caused by 9215a8f722478e5298ccfde66c38c41190ca9748

chrisjurich reviewed (approved):
LGTM!
